shared/utility/math/filter/ParticleFilter.hpp

```diff
@@ -38,242 +74,393 @@ namespace utility::math::filter {
 
-            ParticleList candidate_particles               = ParticleList::Zero(Model::size, particles.cols());
-            candidate_particles.leftCols(particles.cols()) = particles;
-
-            // Resample some rogues
-            for (int i = 0; i < model.getRogueCount(); ++i) {
-                candidate_particles.col(i + model.getParticleCount()) =
-                    particles.col(i) + model.getRogueRange().cwiseProduct(StateVec::Random() * 0.5);
+            // Get the model to make a prediction for each particle
+            // Also get the model to tell us how much a prediction deviates from the observed measurement
+            Eigen::Matrix<MeasurementScalar, S, Eigen::Dynamic> differences(S, particles.cols());
+            for (int i = 0; i < particles.cols(); ++i) {
+                differences.col(i) =
+                    model.difference(model.predict(particles.col(i), std::forward<Args>(params)...), measurement);
             }
 
-            Eigen::Matrix<MeasurementScalar, S, Eigen::Dynamic> differences(S, particles.cols());
-            for (int i = 0; i < candidate_particles.cols(); ++i) {
-                differences.col(i) = model.difference(model.predict(particles.col(i), params...), measurement);
+            // Recalculate the weights of each particle
+            // The weight update factor comes from the density function of a zero mean, measurement_covariance
+            // covariance multivariate normal distribution
+            MultivariateNormal<Scalar, S> multivariate(measurement_covariance);
+            for (int i = 0; i < particles.cols(); ++i) {
+                weights[i] *= multivariate.density(differences.col(i));
             }
 
             // Calculate log probabilities
             Eigen::Matrix<MeasurementScalar, Eigen::Dynamic, 1> logits =
-                -differences.cwiseProduct(measurement_variance.inverse() * differences).colwise().sum();
+                -differences.cwiseProduct(measurement_covariance.inverse() * differences).colwise().sum();
 
-            // Subtract the max log prob for numerical stability and then exponentiate
+            // Subtract the max log probability for numerical stability and then exponentiate
             logits = Eigen::exp(logits.array() - logits.maxCoeff());
 
-            // Resample
-            std::discrete_distribution<> multinomial(logits.data(), logits.data() + logits.size());
-            for (unsigned int i = 0; i < particles.cols(); i++) {
-                particles.col(i) = model.limit(candidate_particles.col(multinomial(rng)));
-            }
+            // Return the mean log probability
             return logits.mean();
         }
 
-        // Take the mean of all particles, pay no attention to the type of the data
-        StateVec get() const {
+        /**
+         * @brief Calculates and returns the mean of all particles.
+         */
+        [[nodiscard]] StateVec get() const {
```

Comment:
I feel like `get()` should be `getMean()` given that the other getter functions are specified